doc/src/manual/networking-and-streams.md

```diff
@@ -312,3 +312,39 @@ resolution:
 julia> getaddrinfo("google.com")
 ip"74.125.226.225"
 ```
+
+## Asynchronous I/O
+All I/O operations exposed by [`Base.read`](@ref) and [`Base.write`](@ref) can be performed
+asynchronously through the use of [coroutines](@ref man-tasks). You can create a new coroutine to
+read from or write to a stream using the [`@async`](@ref) macro:
+
+```julia-repl
+julia> task = @async open("foo.txt", "w") do io
+           write(io, "Hello, World!")
+       end;
+
+julia> wait(task)
+
+julia> readlines("foo.txt")
+1-element Array{String,1}:
+ "Hello, World!"
+```
+
+It's common to run into situations where you want to perform multiple asynchronous operations
+concurrently and wait until they've all completed. You can use the [`@sync`](@ref) macro to cause
+your program to block until all of the coroutines it wraps around have exited:
+
+```julia-repl
+julia> @sync for filename in ("foo.txt", "bar.txt", "baz.txt")
+           @async open(filename, "w") do io
+               write(io, "My name is $(filename)")
```

Comment:
```suggestion
               println("Start writing $filename")
               write(io, "My name is $filename")
               println("Done writing $filename")
```
This seems like it would help demonstrate the asynchronous nature of the process.

Reply:
That seems like a good idea. I've looked at a few different options, and it seems a little difficult to find a good example using file I/O. This socket I/O example is probably fairly reproducible (with connection numbers shuffled around), but a little contrived:

```julia
julia> using Sockets

julia> @sync for i = 1:5
           conn = connect("julialang.org", 80)
           @async begin
               # Write ~32Kb of data, one byte at a time
               for j = 1:2^15
                   write(conn, rand(UInt8, 1))
               end
               println("Closed connection $(i)")
           end
       end
Closed connection 5
Closed connection 2
Closed connection 1
Closed connection 4
Closed connection 3
```

Another (possibly less reproducible) example would be

```julia
julia> using Sockets

julia> @sync for hostname in ("google.com", "github.com", "julialang.org")
           conn = connect(hostname, 80)
           @async begin
               write(conn, "GET / HTTP/1.1\r\nHost:$(hostname)\r\n\r\n")
               readline(conn, keep=true)
               println("Finished connection to $(hostname)")
           end
       end
Finished connection to google.com
Finished connection to julialang.org
Finished connection to github.com
```

Reply:
Yeah, I tried for a while to come up with good examples, but it's surprisingly hard (as you're seeing) to get the I/O subsystem to yield back to the scheduler. The simplest way by far is to call `sleep(1)` but that feels like cheating / too artificial.

Reply:
@JeffBezanson any bright ideas for something simple that will yield in this case?

Reply:
There hasn't been any more activity on this (I'm sure you're all pretty busy!), so I'd like to bring it back to life if possible. Asynchronous I/O is an important topic that's left almost entirely unmentioned in the manual's I/O section. I'd be happy to make any more additions / modifications needed to get this into the documentation.

It feels like the second example _ought_ to be good enough for demonstrating the `@sync` outer block / `@async` inner block. That way the Networking / Streams section will also have one code example explicitly demonstrating asynchronous file I/O, and another explicitly demonstrating asynchronous network I/O.

What else could be changed or added to this section to improve it?